src/Xamarin.Android.Build.Tasks/Tasks/ResolveLibraryProjectImports.cs

```diff
@@ -154,26 +160,95 @@ static string GetTargetAssembly (ITaskItem assemblyName)
 					.Select (a => GetTargetAssembly (a))
 					.Where (a => a != null)
 					.Distinct ()) {
-				foreach (var imp in new string [] {imports_dir, "library_project_imports"}.Distinct ()) {
-					string assemblyIdentName = Path.GetFileNameWithoutExtension (assemblyPath);
-					if (UseShortFileNames) {
-						assemblyIdentName = Xamarin.Android.Tasks.MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly (assemblyIdentName);
-					}
-					string outDirForDll = Path.Combine (OutputImportDirectory, assemblyIdentName);
-					string importsDir = Path.Combine (outDirForDll, imp);
+				string assemblyIdentName = Path.GetFileNameWithoutExtension (assemblyPath);
+				if (UseShortFileNames) {
+					assemblyIdentName = Xamarin.Android.Tasks.MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly (assemblyIdentName);
+				}
+				string outDirForDll = Path.Combine (OutputImportDirectory, assemblyIdentName);
+				string importsDir = Path.Combine (outDirForDll, ImportsDirectory);
+#if SEPARATE_CRUNCH
+				// FIXME: review these binResDir thing and enable this. Eclipse does this.
+				// Enabling these blindly causes build failure on ActionBarSherlock.
+				//string binResDir = Path.Combine (importsDir, "bin", "res");
+				//string binAssemblyDir = Path.Combine (importsDir, "bin", "assets");
+#endif
+				string resDir = Path.Combine (importsDir, "res");
+				string assemblyDir = Path.Combine (importsDir, "assets");
+
+				// Skip already-extracted resources.
+				var stamp = new FileInfo (Path.Combine (outdir.FullName, assemblyIdentName + ".stamp"));
+				if (stamp.Exists && stamp.LastWriteTime > new FileInfo (assemblyPath).LastWriteTime) {
+					Log.LogDebugMessage ("Skipped resource lookup for {0}: extracted files are up to date", assemblyPath);
 #if SEPARATE_CRUNCH
-					// FIXME: review these binResDir thing and enable this. Eclipse does this.
+					// FIXME: review these binResDir/binAssemblyDir thing and enable this. Eclipse does this.
 					// Enabling these blindly causes build failure on ActionBarSherlock.
-					//string binResDir = Path.Combine (importsDir, "bin", "res");
-					//string binAssemblyDir = Path.Combine (importsDir, "bin", "assets");
+					if (Directory.Exists (binResDir))
+						resolvedResourceDirectories.Add (binResDir);
+					if (Directory.Exists (binAssemblyDir))
+						resolvedAssetDirectories.Add (binAssemblyDir);
 #endif
-					string resDir = Path.Combine (importsDir, "res");
-					string assemblyDir = Path.Combine (importsDir, "assets");
+					if (Directory.Exists (resDir))
+						resolvedResourceDirectories.Add (resDir);
+					if (Directory.Exists (assemblyDir))
+						resolvedAssetDirectories.Add (assemblyDir);
+					continue;
+				}
+
+				if (Directory.Exists (outDirForDll))
+					Directory.Delete (outDirForDll, true);
```

Comment:
This presents a theoretical problem -- which presumably won't be all *that* theoretical? -- in which two assemblies have an md2 which is the same value. (This should plausibly be *more* common, since we're only taking the first *5* values in the md2 value! See `MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly()`. That said, I don't see any issues in *my* `$HOME/.local/share/Xamarin`, so maybe conflict isn't *that* likely?)

If we have two directories which hash to the same value, what should be done, and how can it be supported?

(*Can* it be supported?)

For example, assume that assemblies A1 and A2 have names which hash to the same value. The `Directory.Delete()` here will nuke the extraction of one, which will *presumably* result in a later build break, as e.g. some resource is used which doesn't exist.

Alternatively, instead of the `Directory.Delete()`, we could instead extract both files into the same directory. The problem with this approach is deleting entries which no longer exist (bleh). Alternatively, there would be a problem if both assemblies provide a resource of the same name, e.g. `classes.jar` (which iirc is a fairly common name).

Offhand, I'm not immediately sure if this is solvable. At the same time, I'm also not sure if its "fixable", *or* that it even *should* be fixed.

What *might* be a good idea, though, is to *check for it*: we have all of the `Assemblies` of interest, and therefore we can compute the abbreviated md2 for *all* of those assemblies, and check to see if there are *any* conflicts at all:

```csharp
    var haveConflicts = new Dictionary<string, int>();
    foreach (var a in Assemblies) {
        var n = Xamarin.Android.Tasks.MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly (Path.GetFileNameWithoutExtension (a));
        int c;
        if (!haveConflicts.TryGetValue (n, out c)) {
            haveConflicts.Add (n, 1);
        }
        else haveConflicts [n] = c+1;
    }
    if (haveConflicts.Values.Any (v => v > 1))
        Log.LogError ("We have a hash conflict! Please rebuild with `$(UseShortFileNames)`=False.");
```

That's clearly not a great error message, but the *idea* is there.